### ISSUE TITLE
Kubelet secret volume plugin should store data in tmpfs

### ISSUE THREAD

**Issue by pmorie** [GoogleCloudPlatform/kubernetes#4602]
Per #4126 the kubelet secret volume plugin should store data in a tmpfs.  In feedback on #4126, @erictune raised the question of whether the secret volume plugin or the kubelet should handle creating the tmpfs.

Personally I had been in favor of having the kubelet create this and share it via a new method on the `kubelet/volume.Host` interface, like so:

```go
type Host interface {
  // other methods omitted
  GetVolatileStorageDir() string
}
```

My reasoning was that the kubelet will need volatile storage for its own secrets (ie, secrets associated with a `ServiceAccount` that affect kubelet behaviors like `docker pull`), so we could just create a single filesystem and subdivide it for the volume plugin and kubelet both to use.  One advantage to this approach is that cluster operators could size the single volume easily without having to make projections about the relative amount of secret data consumed by pods in volumes versus associated with service accounts associated with pods.

I'm not longer sure this is the best way.  Also, service accounts haven't landed yet, so it's arguable that we shouldn't worry about storage for this flavor of secret yet.

An alternative would be:

1.  Make the secret volume plugin to manage its own tmpfs for now
2.  Defer decision about kubelet-scoped secrets until service accounts land; at that time we could:
    1.  Make a single tmpfs that is used for both kubelet-scoped and pod-scoped secrets
    2.  Make a separate tmpfs for kubelet-scoped secrets

Creating this issue to get some discussion on this matter before I begin the changes for tmpfs in the volume plugin.

**Comment by pmorie** [comment]
@erictune @smarterclayton @EricMountain-1A @csrwng any thoughts?

**Comment by erictune** [comment]
I now agree with you @pmorie 

The kubelet might also want to use a /tmpfs for other volume types, such as an all-memory /emptydir.
I think we may end up with three or more directories in the mid-to-long term: local disk, flash with filesystem, and tmpfs.  One of each.

Since the volume plugin uses a prefix to its path, then sharing should work fine.


I think for now, the cluster operators don't need to specify anything: the volume can grow infinitely.  Eventually, we will need a way to charge disk space, flash space, and tmpfs space of volumes to their pod.  That is not specific to secrets: applies to multiple volume types.   But we are punting on it for now.

So, I say go ahead with a single /tmpfs with no or a very large size limit, created by kubelet.  No adjustment of reported node capacity.

**Comment by pmorie** [comment]
@erictune ack, I will name this tmpfsRootDir in the kubelet for now.  Thinking through changes to the Host interface, do we need a new set of methods scoped to the tmpfs, like so?

```go
type Host interface {
  // other methods omitted

  GetTmpfsPluginDir(plugin string)
  GetPodTmpfsPluginDir(podUid types.UID, pluginName string)
  GetPodTmpfsVolumeDir(podUid types.UID, pluginName, volumeName string)
}
```

**Comment by pmorie** [comment]
@erictune A couple more questions I thought of:

1.  What happens in a tmpfs is already mounted at the tmpfsRootDir?  Should we unmount it?  Go down in flames?
2.  Should all of the get*Dir methods in the kubelet have tmpfs variants?

**Comment by thockin** [comment]
I have now thoroughly reviewed the secrets work (nice work guys, really).

Since I nominally own volumes, let's talk about tmpfs.

I do want to offer a tmpfs EmptyDir.  The obvious thing to do is to add a
tmpfs plugin.  But WAIT.  It's identical to EmptyDir in pretty much every
way except a call to mount() instead of mkdir().  I expect that eventually
I will want a local SSD EmptyDir too - do I again fork it and just hardcode
the concept?  What about when phase change and NVM and whatever new space
age storage tech emerges?

And what about things like the GitRepo?  I want to make that go away, but
the replacement will need similar concepts.  Do I make N copies of that for
disk, SSD, NVM, PCM, quantum, etc?

Stick with me, I am going somewhere, I think.

In a sense, WHAT you store and WHERE you store it are orthogonal.  But not
REALLY because the full matrix doesn't make sense - we don't want to offer
secrets on disk, and what does it even mean to map PD or NFS or iscsi onto
tmpfs?  That's a nonsensical question.

In a sense, what you want to ask for is a QoS of storage or a set of
attributes that must be met.  For example secrets requires a volatile
storage medium, but it doesn't care much about perf.  Someone who cares
about perf might ask for IOPS > x or throughput > y MB/s.  It seems that an
API like:

```json
    source: emptyDir{ requirements: [ "IOPS >= 1000" ]}
```

...is a better way to express performance.  This happens to be similar to
what @markturansky is pushing with pet storage, but I can't help but thing
that is TOO exact, and not good UX.  But if we HAD such an API, I would
argue that secret volumes are essentially hardcoded to "Volatile"
requirements.

So the question I offer is this: Is it worthwhile to take a swing at this
now, and implement secrets in terms of it?  Or should we punt on this and
just make a kubelet-global tmpfs (which we might want for other things
eventually) and then carve slices off of that for secrets?

I think the latter is probably more expedient but messier.  The former is
going to take longer, but has a better chance of lasting, IMO.

And then a second question: Is secrets really just a special case of
GitRepo, which is just a special case of "run this side-car container" ?
In other words, could we get rid of the secrets volume plugin and instead
just run a container which syncs some Secrets objects into files?  The
ServiceAccount tie-in makes me think maybe not, but i wanted to put it on
the table.

Tim

**Comment by pmorie** [comment]
@thockin Good questions, I'm thinking about it.  Agree, what I have now is awfully messy and noisy.

**Comment by pmorie** [comment]
#### Orthogonality of content and underlying storage attributes

> In a sense, WHAT you store and WHERE you store it are orthogonal.  But not
REALLY because the full matrix doesn't make sense - we don't want to offer
secrets on disk, and what does it even mean to map PD or NFS or iscsi onto
tmpfs?  That's a nonsensical question.

@thockin I had a similar thought re: storage type, and agree, the full matrix is not valid.

One question is how the API should be factored - does ability to specify attributes of the backing storage for a volume cut across all volume sources, or is it a function of an individual source?  Either way there will be a need as you say to validate that the requested storage is acceptable to the plugin.

FWIW, I think this would also result in a nicer factoring of the volumes `Host` interface:

```go
type StorageType string

const (
    StorageTypeLocal    StorageType = "Local"
    StorageTypeVolatile StorageType = "Volatile"
    StorageTypeFlash    StorageType = "Flash"
    // FUTURE: add new types
)

type Host interface {
    GetKubeClient() client.Interface

    GetPluginDir(type StorageType, pluginName string) string
    GetPodPluginDir(type StorageType, podUID types.UID, pluginName string) string
    GetPodVolumeDir(type StorageType, podUID types.UID, pluginName, volumeName string) string
}
```

I have more comments but not time to write them atm, will comment with more late.r

**Comment by erictune** [comment]
I agree with what @thockin about where we want to be.
But, I don't think the details of the Host interface have much bearing on how users express their storage requirements to volumes.  So, I don't think we need to tie those two things together.

@pmorie suggestion above about passing StorageType to GetPluginDir seems fine.

**Comment by pmorie** [comment]
@erictune Yeah, they are orthogonal.  For now I'll go with those and we can change other aspects later (either in this iteration or in the future).

**Comment by smarterclayton** [comment]
> On Feb 20, 2015, at 2:27 AM, Tim Hockin <notifications@github.com> wrote:
> 
> I have now thoroughly reviewed the secrets work (nice work guys, really).
> 
> Since I nominally own volumes, let's talk about tmpfs.
> 
> I do want to offer a tmpfs EmptyDir. The obvious thing to do is to add a
> tmpfs plugin. But WAIT. It's identical to EmptyDir in pretty much every
> way except a call to mount() instead of mkdir(). I expect that eventually
> I will want a local SSD EmptyDir too - do I again fork it and just hardcode
> the concept? What about when phase change and NVM and whatever new space
> age storage tech emerges?
> 
> And what about things like the GitRepo? I want to make that go away, but
> the replacement will need similar concepts.  Do I make N copies of that for
> disk, SSD, NVM, PCM, quantum, etc?
> 
> Stick with me, I am going somewhere, I think.
> 
> In a sense, WHAT you store and WHERE you store it are orthogonal. But not
> REALLY because the full matrix doesn't make sense - we don't want to offer
> secrets on disk, and what does it even mean to map PD or NFS or iscsi onto
> tmpfs? That's a nonsensical question.
> 
> In a sense, what you want to ask for is a QoS of storage or a set of
> attributes that must be met. For example secrets requires a volatile
> storage medium, but it doesn't care much about perf. Someone who cares
> about perf might ask for IOPS > x or throughput > y MB/s. It seems that an
> API like:
> 
> ```json
> source: emptyDir{ requirements: [ "IOPS >= 1000" ]}
> ```
> 
> ...is a better way to express performance. This happens to be similar to
> what @markturansky is pushing with pet storage, but I can't help but thing
> that is TOO exact, and not good UX. But if we HAD such an API, I would
> argue that secret volumes are essentially hardcoded to "Volatile"
> requirements.
> 
> So the question I offer is this: Is it worthwhile to take a swing at this
> now, and implement secrets in terms of it? Or should we punt on this and
> just make a kubelet-global tmpfs (which we might want for other things
> eventually) and then carve slices off of that for secrets?
> 
> I think the latter is probably more expedient but messier. The former is
> going to take longer, but has a better chance of lasting, IMO.
> 
> And then a second question: Is secrets really just a special case of
> GitRepo, which is just a special case of "run this side-car container" ?
> In other words, could we get rid of the secrets volume plugin and instead
> just run a container which syncs some Secrets objects into files? The
> ServiceAccount tie-in makes me think maybe not, but i wanted to put it on
> the table.

With secrets, you definitely want to guarantee the secrets are on disk and present before the pod starts.  So that gets into container start ordering - it's definitely a container more like pod-infra than a normal side car.  I also suspect that virtually every pod will have secrets (and many).  So having secrets be a side car entails we would have one side car per pod across the whole cluster.

**Comment by erictune** [comment]
Responding to thockin's comment that maybe we should do per-pod tmpfs.

I think either per-pod or per-node tmpfs can work. Let's build a list of pros and cons together. I'll start.
 - rotating disks can't easily be filesystem-per-pod. Perhaps some advantage to consistency across media types. Result: advantage per-node
 - whatever mechanism we use for resource isolation on rotating disks with filesystems should also work for tmpfs with filesystems. Result: tie
 - I can't see there being much performance difference. Result: tie

**Comment by pmorie** [comment]
> Having secrets be a side car entails we would have one side car per pod across the whole cluster.

I agree with @smarterclayton here. You shouldn't *have* to make a side car container to get secrets in a volume.  There's nothing that says you can't make a sidecar container that gets your secrets from the API server and puts them into your own tmpfs-backed volume.

**Comment by thockin** [comment]
On Fri, Feb 20, 2015 at 11:57 AM, Eric Tune <notifications@github.com> wrote:
>
> Responding to thockin's comment that maybe we should do per-pod tmpfs.
>
> I think either per-pod or per-node tmpfs can work. Let's build a list of pros and cons together. I'll start.
>
> rotating disks can't easily be filesystem-per-pod. Perhaps some advantage to consistency across media types. Result: advantage per-node
> whatever mechanism we use for resource isolation on rotating disks with filesystems should also work for tmpfs with filesystems. Result: tie

Actually, I believe this is not true.  AFAIK traditional quota does
not apply to tmpfs - the only way to limit a tmpfs is to pass a mount
flag.  I'm assuming that this will eventually matter, but maybe not.
As long as you only write to this tmpfs from within your pod it should
be charged to the pod.  Oh, but we don't have pod-scope resources yet.
Oh, and secrets are written by kubelet.  Now that I write it out, I do
think it will matter.

Lastly, if we Greg (another Googler Eric knows) is able to get his
tmpfs-memcg patch, then per-pod matters more.

Tim

**Comment by thockin** [comment]
I'm really torn.  I'm not sure we want kubelet to be in the business of
refreshing your secrets on semi-arbitrary timescales, but I am not sure
it's wrong either.  I can handle it doing a one-time prep, for sure.

If we can formalize side-car containers such that they run BEFORE app
containers, is that really so bad of an approach?

**Comment by smarterclayton** [comment]
Eric and I spent some time talking about this at the f2f but I don't remember the outcome.  I think we said that secrets should propagate as part of rolling update, and that volume status should convey the uid and last resource version of the secret used.  That fits into the broad swath of other behaviors that depend on a roll, and is predictable and consistent.

When per node controllers need to rotate secrets, they still have to have a rotation pattern, so that mechanism would cover that path as well.

**Comment by erictune** [comment]
Yes.  I not longer think there should be changes to the secrets after volume setup time.   I explained why when I closed #4673

**Comment by erictune** [comment]
@thockin I didn't realize quota won't work on tmpfs, so that is good to know.

For the general case of specifying arbitrary media types for arbitrary volumes, I see now that tmpfs per pod helps.

But, for the narrower and more immediate problem of keeping secrets off disk, would it be alright if @pmorie implemented the simpler single tmpfs approach?

**Comment by erictune** [comment]
@pmorie will it work if the secret volume directory is mounted readonly in the container?  That way, kubelet knows the the pod's memory usage is equal to the sum of it's secret sizes.

**Comment by thockin** [comment]
I am very much OK doing a less-than-full implementation.  I'm filing an
issue to capture some of this dicussion

**Comment by pmorie** [comment]
@erictune @thockin Is there any reason to keep this issue open now?  I think we can close it since #5162 and #5167 are both in.
